janusgraph-test/src/test/java/org/janusgraph/graphdb/tinkerpop/optimize/AdjacentVertexHasUniquePropertyOptimizerStrategyTest.java

```diff
@@ -0,0 +1,78 @@
+// Copyright 2020 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.tinkerpop.optimize;
+
+import org.apache.tinkerpop.gremlin.process.traversal.TraversalStrategy;
+import org.apache.tinkerpop.gremlin.process.traversal.step.filter.HasStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.filter.NoneStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.map.NoOpBarrierStep;
+import org.janusgraph.graphdb.configuration.GraphDatabaseConfiguration;
+import org.janusgraph.graphdb.tinkerpop.optimize.strategy.AdjacentVertexHasUniquePropertyOptimizerStrategy;
+import org.junit.jupiter.api.Test;
+
+import static org.janusgraph.graphdb.JanusGraphBaseTest.option;
+import static org.janusgraph.testutil.JanusGraphAssert.assertNumStep;
+import static org.janusgraph.testutil.JanusGraphAssert.assertSameResultWithOptimizations;
+
+/**
+ * @author Florian Grieskamp (Florian.Grieskamp@gdata.de)
+ */
+public class AdjacentVertexHasUniquePropertyOptimizerStrategyTest extends OptimizerStrategyTest {
+
+    private static TraversalStrategy optimizer = AdjacentVertexHasUniquePropertyOptimizerStrategy.instance();
+
+    @Test
+    public void testWithAndWithoutStrategy() {
+        makeSampleGraph();
+        assertNumStep(1, 0, g.V(sv[0]).out().has("uniqueId", 0), HasStep.class);
+        assertNumStep(1, 1, graph.traversal().withoutStrategies(AdjacentVertexHasUniquePropertyOptimizerStrategy.class).V(sv[0]).out().has("uniqueId", 0), HasStep.class);
+    }
+
+    @Test
+    public void testWithBackendAccess() {
```

Comment:
What about the case you mentioned in the https://github.com/JanusGraph/janusgraph/pull/2189#issue-472481806, i.e.,

g.V().has('user', 'name', 'alice').out('knows').as('x').in('knows').has('user', 'name', 'bob').select('x')

Reply:
That is an equivalent example which is just more readable because of the given names. I don't even remember why I used step labels for that example, as they only add complexity. Probably did so in order to point out that this part can be nested in arbitrary traversals.